From 56ae80fe56f7dc84a51beaf7c6f2f0efce7acdc5 Mon Sep 17 00:00:00 2001
From: Harald Sitter <sitter@kde.org>
Date: Mon, 2 Jan 2023 12:38:02 +0100
Subject: [PATCH 079/144] correctly set up ref counting in
 QThreadPool::tryStart(std::function)

this function was inconsistent with the QThreadPool::tryStart(QRunnable)
overload, where ref counting does get set up correctly. the lack of ref
counting setup would later cause trouble because we assert the ref
counting state all over QThreadPool. to prevent failed assertions we
correctly set up ref counting now.

this change is not applicable to Qt6 because the ref counting has
already been removed there:
https://github.com/qt/qtbase/commit/ecfda98d1f91c6a7da0d89826558d856cd88e670

BUG: 449688
---
 src/corelib/thread/qthreadpool.cpp | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/corelib/thread/qthreadpool.cpp b/src/corelib/thread/qthreadpool.cpp
index cdeb077233..5b1e4f4c33 100644
--- a/src/corelib/thread/qthreadpool.cpp
+++ b/src/corelib/thread/qthreadpool.cpp
@@ -602,8 +602,12 @@ bool QThreadPool::tryStart(std::function<void()> functionToRun)
         return false;
 
     QRunnable *runnable = QRunnable::create(std::move(functionToRun));
+    Q_ASSERT(runnable->ref == 0);
+    ++runnable->ref;
     if (d->tryStart(runnable))
         return true;
+    --runnable->ref;
+    Q_ASSERT(runnable->ref == 0);
     delete runnable;
     return false;
 }
-- 
2.43.0

